### DISCUSSION TITLE
http.POST streaming the payload

### DISCUSSION THREAD

**Discussion by johnofleek** [earlephilhower/arduino-pico#1601] [Q&A]
I have a large payload I'd like to http.POST

`
int httpCode = http.POST(binaryPayload, binaryPayload_size);
`
Fails above a 0x4000 byte (16384) size payload

**Edit:**
I think that the underlying POST code limits the payload sent to IP to 128 bytes max ? Therefore I'm guessing that my issue isn't related to the payload size but is cause by something else ? From HTTPClient.cpp showing I think the code that sends 128 bytes at a time into TCP/IP

![image](https://github.com/earlephilhower/arduino-pico/assets/5321355/03d742fa-4c5d-462c-8800-23895b321ee0)

**Edit2**
Short story - enabling debugging didn't help so far

![image](https://github.com/earlephilhower/arduino-pico/assets/5321355/d7a0b014-6a52-4b2a-950f-4f0cd79ecca3)

I did hack this bit of WiFiClientSecureBearSSL.cpp because I couldn't quickly work out a proper solution - but the issues aren't in that part anyway

![image](https://github.com/earlephilhower/arduino-pico/assets/5321355/6b24dda4-401c-43c7-bb7d-54aff3d535f8)

**Comment by maxgerhardt** [comment]
The `sendAll()` function transports the data in **chunks** of maximum size 128 bytes to the "client". I think 16K payload limitation might come from the used firmware in the WiFi chip, the Cypress CW.. something on the Pico W  (I assume that's what you're using)., that's where the payload will be written to in the most immediate path.

**Reply by earlephilhower** [reply]
It looks like the code is actually doing HTTPS and not HTTP.  In that case, 16K is the TLS record size, meaning the core will buffer up 16K of writes (at 128b a pop) into that buffer, encrypt, and then transmit it using TCP.  There may be some HTTP-specific things (i.e. you might need to use HTTP chunked encoding) when multiple TCP packets are sent in via a POST.

The only place I think HTTPClient can return that error is here:
https://github.com/earlephilhower/arduino-pico/blob/67a07edcccb954c7ef8938ff64733f4a89eff893/libraries/HTTPClient/src/HTTPClient.cpp#L598-L601

The `StreamConstPtr` is a hack I pulled in to allow the ESP8266 HTTPClient to run (the 8266 rewrote the whole Stream class to be lower memory but incompatible with plain Arduino). 
https://github.com/earlephilhower/arduino-pico/blob/67a07edcccb954c7ef8938ff64733f4a89eff893/libraries/HTTPClient/src/HTTPClient.cpp#L85-L101

So this could fail if a) 5s have passed since the start of the transmission, or b) the `WiFiClientSecure::write` returns 0.

It would be instructive if you could make a small MCVE that posts against some public webserver or a simple python one to repro and get to the bottom of it.
